### PR TITLE
Support option to disable DNSSEC validation

### LINKED ISSUE #37: Support option to disable DNSSEC verification
I'm using dohdec as a resolver in a [DNSSEC library](https://github.com/relaycorp/dnssec-js), which means Google/Cloudflare/etc shouldn't be doing the validation. The problem is that, when DNSSEC fails, the server doesn't return any records.

Would you accept a PR that implemented an option to enable the [`cd` flag](https://developers.google.com/speed/public-dns/docs/doh/json#supported_parameters)?

### PR DESCRIPTION
**Pull request by gnarea**
This enables the `cd` flag when DNSSEC records are requested, but we want to disable DNSSEC validation on the resolver.

I created this test file to test the changes against real resolvers (but didn't commit it):

```js
// pkg/dohdec/test/real.ava.js

import {DNSoverHTTPS} from '../lib/doh.js'
import {DNSoverTLS} from '../lib/index.js'
// eslint-disable-next-line node/no-missing-import
import test from 'ava'

test('DoH - no dnssec', async t => {
  const doh = new DNSoverHTTPS()
  const r = await doh.lookup('ietf.org', {
    json: false,
    dnssec: false,
  })
  t.truthy(r)

  t.false(r.answers.some(a => a.type === 'RRSIG'))
  t.false(r.flag_cd)
})

test('DoH - dnssec', async t => {
  const doh = new DNSoverHTTPS()
  const r = await doh.lookup('ietf.org', {
    json: false,
    dnssec: true,
  })
  t.truthy(r)

  t.true(r.answers.some(a => a.type === 'RRSIG'))
  t.false(r.flag_cd)
})

test('DoH - dnssec failed', async t => {
  const doh = new DNSoverHTTPS()
  const r = await doh.lookup('dnssec-failed.org', {
    json: false,
    dnssec: true,
  })
  t.truthy(r)

  t.is(r.rcode, 'SERVFAIL')
  t.is(r.answers.length, 0)
  t.false(r.flag_cd)
})

test('DoH - dnssec failed with cd=1', async t => {
  const doh = new DNSoverHTTPS()
  const r = await doh.lookup('dnssec-failed.org', {
    json: false,
    dnssec: true,
    dnssecCd: true,
  })
  t.truthy(r)

  t.is(r.rcode, 'NOERROR')
  t.true(r.answers.some(a => a.type === 'RRSIG'))
  t.true(r.flag_cd)
})

test('DoT - no dnssec', async t => {
  const dot = new DNSoverTLS()
  const r = await dot.lookup('ietf.org', {
    json: false,
    dnssec: false,
  })
  t.truthy(r)

  t.false(r.answers.some(a => a.type === 'RRSIG'))
  t.false(r.flag_cd)
})

test('DoT - dnssec', async t => {
  const dot = new DNSoverTLS()
  const r = await dot.lookup('ietf.org', {
    json: false,
    dnssec: true,
  })
  t.truthy(r)

  t.true(r.answers.some(a => a.type === 'RRSIG'))
  t.false(r.flag_cd)
})

test('DoT - dnssec failed', async t => {
  const dot = new DNSoverTLS()
  const r = await dot.lookup('dnssec-failed.org', {
    json: false,
    dnssec: true,
  })
  t.truthy(r)

  t.is(r.rcode, 'SERVFAIL')
  t.is(r.answers.length, 0)
  t.false(r.flag_cd)
})

```

Fixes #37

### CODE REVIEW

File: pkg/dohdec/types/dot.d.ts
Comment (gnarea) [inline pkg/dohdec/types/dot.d.ts:164]: This change is unrelated strictly speaking, but I noticed it works with JSON too now so I thought I might as well update it whilst I'm there.

Comment (gnarea) [reply pkg/dohdec/types/dot.d.ts]: Never mind, it does require `json=false`.

**Comment by gnarea** [comment]
I spotted a couple of bugs whilst testing the PR but now everything looks good and I won't be making further changes until you provide some feedback.

File: pkg/dohdec/test/dnsUtils.ava.js
Comment (hildjj) [inline pkg/dohdec/test/dnsUtils.ava.js:74]: does `{dnssec: false, dnssecCd: true}` make sense?  If not, should it be an error?

File: pkg/dohdec/types/doh.d.ts
Comment (hildjj) [inline pkg/dohdec/types/doh.d.ts:144]: The rest of these options have more friendly names.  I wonder if we should spell out `dnssecCheckingDisabled`?

**Review by hildjj — CHANGES_REQUESTED** [review]
This looks pretty close.  I had a few minor questions that aren't blockers.
I'd like two changes, please:
- Add yourself to a new contributors section in the package.json
- Add a corresponding CLI option

Only the first is mandatory, I can hack the CLI option in if you don't have time.

File: pkg/dohdec/test/fixtures/doh.ava.js.json
Comment (gnarea) [reply pkg/dohdec/test/fixtures/doh.ava.js.json:354]: I did it by hand. I didn't realise the file was generated.

How can I record it? I just came across `ava-playback` but that doesn't seem to be what you're using (I can't find references to `ava-playback` in the source code).

File: pkg/dohdec/test/dnsUtils.ava.js
Comment (gnarea) [reply pkg/dohdec/test/dnsUtils.ava.js:74]: Good question. I can't find the answer in the DNSSEC RFCs, but that combination works OK if we set `cd=1` with that combination and run the following test:

```js
test('DoH - dnssec disabled with cd=1', async t => {
  const doh = new DNSoverHTTPS()
  const r = await doh.lookup('dnssec-failed.org', {
    json: false,
    dnssec: false,
    dnssecCd: true,
  })
  t.truthy(r)

  t.is(r.rcode, 'NOERROR')
  t.false(r.answers.some(a => a.type === 'RRSIG'))
  t.true(r.flag_cd)
})
```

So I'm going to change the code to set `cd=1` solely based on the value of `dnssecCheckingDisabled` (fka `dnssecCd`), regardless of the value of `dnssec`.

File: pkg/dohdec/test/fixtures/doh.ava.js.json
Comment (hildjj) [reply pkg/dohdec/test/fixtures/doh.ava.js.json:354]: I should add this to the README, sorry.  I'm using [nock](https://github.com/nock/nock#readme) to mock the HTTP requests.  In my history file, I've got `NOCK_BACK_MODE=record npx ava test/doh.ava.js`, see if that works?

File: pkg/dohdec-cli/lib/cli.js
Comment (gnarea) [inline pkg/dohdec-cli/lib/cli.js:89]: I'm not super happy with the `-k` (taken from `chec_k_ing`) but couldn't find another relevant letter that wasn't already taken.

File: pkg/dohdec/test/fixtures/doh.ava.js.json
Comment (gnarea) [reply pkg/dohdec/test/fixtures/doh.ava.js.json:354]: No worries. Thanks!

Running that command runs the tests and they all pass, but `doh.ava.js.json` isn't changed.

If I revert my changes to `doh.ava.js.json` and run `NOCK_BACK_MODE=record npx ava test/doh.ava.js`, my new test in `doh.ava.js` fails as follows:

```
Nock: No match for request {␊
      "method": "GET",␊
      "url": "https://cloudflare-dns.com/dns-query?name=ietf.org&type=MX&do=1&cd=1&random_padding=8FnucR2WQ.RXKio0uPpinbMsspJ1NdVlk~0CVdn-iwll",␊
      "headers": {␊
        "user-agent": "dohdec v5.0.3",␊
        "accept": "application/dns-json",␊
        "accept-encoding": "gzip, deflate, br"␊
      }
```

So I guess it isn't actually recording, right? I wonder whether there's something else I need to do but I don't know what that may be, having checked [the docs](https://github.com/nock/nock#modes).

File: pkg/dohdec-cli/lib/cli.js
Comment (hildjj) [inline pkg/dohdec-cli/lib/cli.js:184]: the `dnssecCd` doesn't look right there?

Comment (gnarea) [reply pkg/dohdec-cli/lib/cli.js:184]: Good catch! Will fix it

File: pkg/dohdec/test/fixtures/doh.ava.js.json
Comment (hildjj) [reply pkg/dohdec/test/fixtures/doh.ava.js.json:354]: Let's not block on this.  Let's get it landed, and I'll fiddle with it after I update all the dependencies and fix whatever stupid thing node 19 broke.

File: pkg/dohdec/test/dnsUtils.ava.js
Comment (hildjj) [reply pkg/dohdec/test/dnsUtils.ava.js:74]: you haven't made this change yet, right?

File: pkg/dohdec-cli/lib/cli.js
Comment (gnarea) [reply pkg/dohdec-cli/lib/cli.js]: Fixed:

```
$ node bin/dohdec.js --dnssec --dnssecCheckingDisabled
domain (rrtype)> dnssec-failed.org
[
  {
    name: 'dnssec-failed.org',
    type: 1,
    TTL: 300,
    data: '96.99.227.255'
  },
  {
    name: 'dnssec-failed.org',
    type: 46,
    TTL: 300,
    data: 'A RSASHA1 2 300 1670251862 1668782762 44973 dnssec-failed.org. FlsTSg98nrXHmX5TLfORtA5r1DzyWojyTc6Dl81X4ssm6R5LKwUEin1D/8/MJoGRe7QKFpdL3VB//wDoilUiCUGoZDAqT/uF2Mi89PkOzyfIAy3Uv59nWvZJSfI369So/E991zLmZDekXjNTrJV43UtWhw5Wn2gim7LlX/Fsiog='
  }
]
```

File: pkg/dohdec/test/fixtures/doh.ava.js.json
Comment (gnarea) [reply pkg/dohdec/test/fixtures/doh.ava.js.json:354]: Cool, thanks!

File: pkg/dohdec/test/dnsUtils.ava.js
Comment (gnarea) [reply pkg/dohdec/test/dnsUtils.ava.js:74]: I did: https://github.com/hildjj/dohdec/pull/38/commits/acd49694a83825a461bdff55e4a4a63ca7a4bbef

Comment (hildjj) [reply pkg/dohdec/test/dnsUtils.ava.js:74]: Sorry, i must have been looking at an outdated thing.  I see it now!

**Review by hildjj — APPROVED** [review]

**Comment by hildjj** [comment]
OK, this looks ready to go to me.  There's a solid 10% chance this is going to break GHA when it lands, because I hadn't updated the action definition to run on pull requests to `main` instead of `master`, but I'll deal with that in my update PR that will come after this.  Anything else before I merge?

**Comment by gnarea** [comment]
Thank you! I think we're good to go! 🚀

**Comment by gnarea** [comment]
Hi @hildjj 👋🏾 Can you please release this change to NPM? I've been relying on [the Git tag](https://github.com/relaycorp/dnssec-js/blob/ce5c9a52b759cc7a5fc7b59c7b673da14cefde1a/package.json#L47) but [that's now breaking CI for me](https://github.com/relaycorp/dnssec-js/pull/276#issuecomment-2323405722). TIA! 🙇🏾

**Comment by hildjj** [comment]
I'll look at this first thing Tuesday.  I think this is blocked on https://github.com/hildjj/mock-tls-server/issues/2, so I either have to fix that or rewrite the tests.

**Comment by hildjj** [comment]
OK, I think I'm ready to cut a major release, but I'm too tired today to re-learn lerna.  Release probably will happen tomorrow morning MDT.

**Comment by hildjj** [comment]
Version 6.0.0 just released.  Please check to make sure I got it right... I had to do a bunch of manual hacking to get the release to publish from GHA with provenance information correct.

**Comment by gnarea** [comment]
Thanks @hildjj! [I tried upgrading](https://github.com/relaycorp/dnssec-js/pull/277), but it looks like the changes I made to `pkg/dohdec/types/doh.d.ts` in this PR didn't make it to NPM:

```
src/integration_tests/retrieval.test.ts:38:9 - error TS2345: Argument of type '{ rrtype: "A" | "NS" | "MD" | "MF" | "CNAME" | "SOA" | "MB" | "MG" | "MR" | "NULL" | "WKS" | "PTR" | "HINFO" | "MINFO" | "MX" | "TXT" | "RP" | "AFSDB" | "X25" | "ISDN" | "RT" | "NSAP" | ... 66 more ... | "DLV"; json: false; decode: false; dnssec: true; dnssecCheckingDisabled: boolean; }' is not assignable to parameter of type 'string | DOH_LookupOptions | undefined'.
  Object literal may only specify known properties, and 'dnssecCheckingDisabled' does not exist in type 'DOH_LookupOptions'.

38         dnssecCheckingDisabled: true,
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in src/integration_tests/retrieval.test.ts:38
```

**Comment by hildjj** [comment]
The .d.ts files are generated, and aren't even checked into git anymore.  I'll see if I can find the right place to change the source.

**Comment by hildjj** [comment]
It looks like you only added it on `getJSON()`, and didn't get it passed all the way through the types when you use `lookup()`.

WillFix.

**Pull request merged**
